pinot-broker/src/main/java/com/linkedin/pinot/broker/routing/RoutingTableBuilderFactory.java

```diff
@@ -99,18 +109,34 @@ public RoutingTableBuilder createRoutingTableBuilder(TableConfig tableConfig) {
       case PartitionAwareOffline:
         // In order to use the partition aware routing for offline table, we assume the following conditions.
         // 1. The replica group aware segment assignment strategy is used.
-        // 2. Table config contains the replica group segment assignment related configurations.
-        // 3. The table push type is not 'refresh'.
-        String segmentAssignmentStrategy = tableConfig.getValidationConfig().getSegmentAssignmentStrategy();
+        // 2. The table push type is not 'refresh'.
         SegmentsValidationAndRetentionConfig validationConfig = tableConfig.getValidationConfig();
+        String segmentAssignmentStrategy = validationConfig.getSegmentAssignmentStrategy();
+
         if (CommonConstants.Helix.DataSource.SegmentAssignmentStrategyType.valueOf(segmentAssignmentStrategy)
-            != CommonConstants.Helix.DataSource.SegmentAssignmentStrategyType.ReplicaGroupSegmentAssignmentStrategy ||
-            validationConfig.getReplicaGroupStrategyConfig() == null ||
-            (validationConfig.getSegmentPushType() != null && validationConfig.getSegmentPushType()
+            == CommonConstants.Helix.DataSource.SegmentAssignmentStrategyType.ReplicaGroupSegmentAssignmentStrategy &&
+            validationConfig.getReplicaGroupStrategyConfig() != null &&
+            (validationConfig.getSegmentPushType() != null && !validationConfig.getSegmentPushType()
                 .equalsIgnoreCase("REFRESH"))) {
+          builder = new PartitionAwareOfflineRoutingTableBuilder();
+        } else {
           builder = new DefaultOfflineRoutingTableBuilder();
+        }
+        break;
+      case PartitionAwareRealtime:
+        // In order to use the partition aware routing for realtime table, we assume that the following.
+        // 1. The table push type is not 'refresh'
```

Comment:
I was not aware that the realtime segments do not care about the push type. I removed this check.